### PR TITLE
Added new global code placement transformation.

### PR DESCRIPTION
**Pull request by m4rs-mt**
This PR adds a new global (and aggressive 🚀) code transformation that reorders all operations according to their uses. This minimizes the live value ranges and reduces register consumption. To get the maximum performance out of the optimization pipeline, this PR adds another run of the LICM transformation to the pipeline.

### CODE REVIEW

File: Src/ILGPU/IR/Values/SideEffectValue.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+{
+    /// <summary>
+    /// Represents an abstract value with side effects.
+    /// </summary>
+    public abstract class SideEffectValue : Value
+    {
+        #region Instance
+
+        /// <summary>
+        /// Constructs a new memory value.
```
Comment (MoFtZ) [inline Src/ILGPU/IR/Values/SideEffectValue.cs:24]: MINOR: Incorrect comment - is this always a memory value?

File: Src/ILGPU/IR/Values/SideEffectValue.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        /// </summary>
+        /// <param name="initializer">The value initializer.</param>
+        internal SideEffectValue(in ValueInitializer initializer)
+            : base(initializer)
+        { }
+
+        /// <summary>
+        /// Constructs a new memory value.
```
Comment (MoFtZ) [inline Src/ILGPU/IR/Values/SideEffectValue.cs:32]: MINOR: Incorrect comment?

File: Src/ILGPU/IR/Values/SideEffectValue.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        /// </summary>
+        /// <param name="initializer">The value initializer.</param>
+        internal ControlFlowValue(in ValueInitializer initializer)
+            : base(initializer)
+        { }
+
+        /// <summary>
+        /// Constructs a new memory value.
```
Comment (MoFtZ) [inline Src/ILGPU/IR/Values/SideEffectValue.cs:59]: MINOR: Incorrect comment?

File: Src/ILGPU/IR/Transformations/CodePlacement.cs [hunk 4]
```diff
@@ -0,0 +1,876 @@
...
+
+            /// <summary>
+            /// Setups the value insert position for the given block builder.
+            /// </summary>
+            /// <param name="builder">The current block builder.</param>
+            void SetupInsertPosition(BasicBlock.Builder builder);
+        }
+
+        /// <summary>
+        /// Appends values by inserting them behind a all phi values.
```
Comment (MoFtZ) [inline Src/ILGPU/IR/Transformations/CodePlacement.cs:320]: MINOR: Typo.
```suggestion
        /// Appends values by inserting them behind all phi values.
```

File: Src/ILGPU/IR/Transformations/CodePlacement.cs [hunk 5]
```diff
@@ -0,0 +1,876 @@
...
+                // Gather all side effect values in the order of their appearance
+                foreach (var block in blocks)
+                {
+                    foreach (Value value in block)
+                    {
+                        if (!(value is SideEffectValue sev))
+                            continue;
+
+                        valueBlocks[sev] = value.BasicBlock;
+                        if (value is MemoryValue memoryValue)
```
Comment (MoFtZ) [inline Src/ILGPU/IR/Transformations/CodePlacement.cs:434]: QUESTION: Is this MemoryValue correct? Should it be SideEffectValue?
```suggestion
                        if (value is SideEffectValue sideEffectValue)
```

File: Src/ILGPU/IR/Transformations/CodePlacement.cs [hunk 6]
```diff
@@ -0,0 +1,876 @@
...
+
+                // Sweep upwards or downwards and check whether we might break the program
+                // semantics by violating the order of load/store operations
+                bool skipLoads = memoryValue is Load;
+                int increment = startIndex > valueIndex ? -1 : 1;
+                for (int i = startIndex + increment;
+                    lower && i > valueIndex || i < valueIndex;
+                    i += increment)
+                {
+                    // Ignore loads?
```
[the remark this answers is not in the document]
Comment (m4rs-mt) [reply Src/ILGPU/IR/Transformations/CodePlacement.cs:540]: This is an oversight...

File: Src/ILGPU/IR/Transformations/CodePlacement.cs (hunk 5 above)
Comment (m4rs-mt) [reply Src/ILGPU/IR/Transformations/CodePlacement.cs:434]: This should be correct as `MemoryValues` have to appear in a pre-defined order (in order to preserve read-write dependencies). Other values with side effects can depend on the control flow itself (e.g. a `WarpShuffle` value). However, they can be freely moved around as long as they do not end up in a block that violates their "current" control-flow preconditions.

**Review by MoFtZ — APPROVED** [review]

**Review by MoFtZ — CHANGES_REQUESTED** [review]
Looks like some review changes were not made.

**Review by MoFtZ — APPROVED** [review]

**Pull request merged**